TDataFrame.hxx

```diff
@@ -338,37 +338,48 @@ enum class EActionType : short {
    kMean
 };
 
+template<typename Derived> class TDataFrameInterface;
+
+// A proxy for TDataFrame{Filter,Branch}. De facto it is a workaround for a smart reference.
+template<typename Derived>
+class TDataFrameProxy : public TDataFrameInterface<Derived> {
+   Derived *fDerivedPtr;
+public:
+   TDataFrameProxy(TDataFrameInterface<Derived> *iface) : TDataFrameInterface<Derived>(iface), fDerivedPtr(static_cast<Derived*>(iface)) { }
+   TDataFrame& GetDataFrame() const {return fDerivedPtr->GetDataFrame();}
+};
 
 // this class provides a common public interface to TDataFrame and TDataFrameFilter
 // it contains the Filter call and all action calls
 template<typename Derived>
 class TDataFrameInterface {
 public:
    TDataFrameInterface() : fDerivedPtr(static_cast<Derived*>(this)) { }
+   TDataFrameInterface(TDataFrameInterface<Derived>* derived) : fDerivedPtr(static_cast<Derived*>(derived)) { }
    virtual ~TDataFrameInterface() { }
 
    template<typename F>
-   auto Filter(F f, const BranchList& bl = {}) -> TDataFrameFilter<F, Derived>& {
+   auto Filter(F f, const BranchList& bl = {}) -> TDataFrameProxy<TDataFrameFilter<F, Derived>> {
       ::CheckFilter(f);
       const BranchList& defBl = fDerivedPtr->GetDataFrame().GetDefaultBranches();
       const BranchList& actualBl = ::PickBranchList(f, bl, defBl);
       using DFF = TDataFrameFilter<F, Derived>;
       auto FilterPtr = std::unique_ptr<DFF>(new DFF(f, actualBl, *fDerivedPtr));
-      auto& FilterRef = *FilterPtr;
+      TDataFrameProxy<DFF> dffProxy(FilterPtr.get());
```

Comment:
By concurrency, do you mean ownership? If yes, I agree (see last line of the commit comment https://github.com/bluehood/tdataframe/pull/50/commits/39dfc3dfa583adf62286c07fc9bc727c720dc36e). If not, I don't get the comment :)
My thoughts about possible implementations, both come with a price tag:
1) Have a smart pointer instead of the proxy. I am not enthusiastic about this solution since this will prevent the usage of "." to chain operations, which is quite nice and usable.
2) Rely on a solution implementing the CRTP. The price would be to replicate the methods of the TDataFrameInterface.